src/gui/sdlmain.cpp

```diff
@@ -3033,59 +3019,63 @@ static SDL_Point clamp_to_minimum_window_dimensions(SDL_Point size)
+static void setup_viewport_resolution_from_conf(const std::string &viewport_resolution_val)
 {
-	sdl.use_max_resolution = false;
-	sdl.max_resolution = {-1, -1};
+	sdl.use_viewport_limits = false;
+	sdl.viewport_resolution = {-1, -1};
 
 	// TODO: Deprecate SURFACE output and remove this.
 	if (sdl.desktop.want_type == SCREEN_SURFACE)
 		return;
 
-	if (max_resolution_val == "auto")
+	constexpr auto default_val = "fit";
+	if (viewport_resolution_val == default_val)
 		return;
 
-	const auto desktop = get_desktop_resolution();
-	int w, h;
-	auto was_parsed = sscanf(max_resolution_val.c_str(), "%dx%d", &w, &h) == 2;
+	int w = 0;
+	int h = 0;
+	int p = 0;
+	const auto was_parsed = sscanf(viewport_resolution_val.c_str(), "%dx%d", &w, &h) == 2 ||
+	                        sscanf(viewport_resolution_val.c_str(), "%d%%", &p) == 1;
+
 	if (!was_parsed) {
-		LOG_WARNING("DISPLAY: Requested max_resolution '%s' was not in WxH format, resolution restriction is disabled",
-		            max_resolution_val.c_str());
+		LOG_WARNING("DISPLAY: Requested viewport_resolution '%s' was not in WxH"
+		            " or N%% format, using the default setting ('%s') instead",
+		            viewport_resolution_val.c_str(), default_val);
 		return;
 	}
 
-	const bool is_out_of_bounds = w <= 0 || w > desktop.w || h <= 0 ||
-	                              h > desktop.h;
+	const auto desktop = get_desktop_resolution();
+	const bool is_out_of_bounds = (w <= 0 || w > desktop.w || h <= 0 ||
+	                               h > desktop.h) &&
+	                              (p <= 0 || p > 100);
 	if (is_out_of_bounds) {
-		LOG_WARNING("DISPLAY: Requested max_resolution '%dx%d' is outside the bounds of the desktop '%dx%d', "
-		            "resolution restriction is disabled",
-		            w, h, desktop.w, desktop.h);
+		LOG_WARNING("DISPLAY: Requested viewport_resolution of '%s' is outside"
+		            " the desktop '%dx%d' bounds or the 1-100%% range, "
+		            " using the default setting ('%s') instead",
+		            viewport_resolution_val.c_str(), desktop.w,
+		            desktop.h, default_val);
 		return;
 	}
 
-	auto coarse_size = clamp_to_minimum_window_dimensions({w, h});
+	sdl.use_viewport_limits = true;
+	if (p > 0) {
+		sdl.viewport_resolution.x = ceil_sdivide(desktop.w * p, 100);
+		sdl.viewport_resolution.y = ceil_sdivide(desktop.h * p, 100);
+		LOG_MSG("DISPLAY: Limiting viewport resolution to %d%% (%dx%d) of the desktop",
+		        p, sdl.viewport_resolution.x, sdl.viewport_resolution.y);
 
-	SDL_Point refined_size;
-	if (sdl.scaling_mode == SCALING_MODE::PERFECT) {
-		// Keep requested max_resolution in pixel-perfect modes; refining it
-		// would be too counterintuitive
-		refined_size = coarse_size;
 	} else {
-		refined_size = refine_window_size(coarse_size, sdl.scaling_mode,
-		                                  wants_stretched_pixels());
+		sdl.viewport_resolution = {w, h};
```

Comment:
Was removing the special handling for pixel-perfect modes intentional? I guess in one way it's logical not to have this special handling, on the other hand I found PP being applied on top of the viewport resolution restriction a bit counterintuitive.

Reply:
It now actually behaves like it did before for pixel perfect mode (which was: not doing any refining of the viewport dimensions when in pixel-perfect mode).

 But the PR does remove the special handling for non-PP modes; so now the viewport is exactly as requested.